src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,153 @@
+
+So, when a transaction `T1` (either reading or writing) finds that a row `R1` has a lock which belongs to an earlier transaction `T0`, it, in fact, doesn't simply rollback itself immediately. However, it checks the state of the lock's primary.
+
+* If the primary lock disappeared but there's a record `data @ T0.start_ts` in `write` column, it means that `T0` has been successfully committed. Then the row `R1`'s stale lock can also be committed. Usually we call this `rolling forward`. After that, the new transaction `T1` can continue.
+* If the primary lock disappeared with nothing left, it means the transaction has been rolled back. Then the row `R1`'s stale lock should also be rolled back. Then `T1` can continue.
+* If the primary lock exists but it's too old (we can set a TTL to the lock), we can believe that the transaction crashed before committing or rolling back. We can rollback it and continue `T1`.
+* Otherwise, we consider transaction `T0` is still running. Then `T1` can rollback itself, or another choice, try to wait for a while to see whether `T0` will be committed before `T1.start_ts`.
+
+## Percolator in TiKV
+
+TiKV is a key-value storage, and each key-value pair can be regarded as a row in Percolator.
+
+TiKV internally uses RocksDB, a key-value storage engine library, to persist data to disk. RocksDB's atomic write batch and TiKV's transaction scheduler make it atomic to read and write a single user key, which is a requirement of Percolator.
+
+RocksDB provides a feature named *Column Family* (we will abbreviate it as *CF*). An instance of RocksDB may have multiple CFs, and each CF is a separated key namespace and has its own LSM-Tree. However different CFs in the same RocksDB instance uses a common WAL, providing the ability to write to different CFs atomically.
+
+We divide a RocksDB to three CFs: `CF_DEFAULT`, `CF_LOCK` and `CF_WRITE`, which corresponds to Percolator's `data` column, `lock` column and `write` column respectively. There's an extra CF named `CF_RAFT` used to save some metadata of Raft, but that's out of our topic now. The `notify` and `ack_O` columns are not present in TiKV, because for now TiKV doesn't need the ability of incremental processing.
+
+Then, we need to represent different versions of a key. We can simply compound a key and a timestamp as an internal key, and use the internal key in RocksDB. But since a key can have at most one lock at a time, so we don't need to add a timestamp to the key in `CF_LOCK`. Then the contents of each CFs are:
+
+* `CF_DEFAULT`: `(key, start_ts)` -> `value`
+* `CF_LOCK`: `key` -> `lock_info`
+* `CF_WRITE`: `(key, commit_ts)` -> `write_info`
+
+Our technical to compose user keys and timestamps together is:
+
+1. Encode the user key to [memcomparable](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format)
+2. Bitwise invert the timestamp (which is an unsigned int64) and encode it to big-endian bytes
```

Comment:
```suggestion
2. Bitwise invert the timestamp (an unsigned Int64) and encode it into big-endian bytes.
```

Reply:
I think the term "int64" doesn't need to be capitalized

Reply:
Searched on google and get both instances. So I guess it is fine to not capitalize it.

Reply:
Searched on google and got both instances of capitalized and non-capitalized. So I guess it's fine to leave as is.